### DISCUSSION TITLE
problems with libimfort.a

### DISCUSSION THREAD

**Discussion by jeromeorosz** [iraf-community/iraf#252] [Q&A] [answered]
Hi All,

Some 20 years ago, I was able to write Fortran codes using IRAF libraries and PGPLOT libraries to read in a FITS image and make a finding chart.   Years later, I can still deal with the PGPLOT libraries.  However, I am having trouble with the IRAF libraries for 2.17.    Excluding the PGPLOT libraries, here is the compile command:
```shell
gfortran -o plotPSR plotPSR.for libimfort.a libsys.a libvops.a libos.a libf2c.a 
```
The code compiles just find.  However, the code won't run properly.  First of all, it won't accept the ".fits" extension for the images.  I used IRAF and made a copy in ".imh" format.  The code was able to find the image, and apparently open it.  However, I don't think the image was read correctly.  Here is the first part of the code (which ran some 20 years ago):

```Fortran
      real            rpix(2048),rbright(2048,2048),sfred
      real            trans(6),xline(2),yline(2)
      integer*2       spix(4096),ofred,sbright(2048,2048)
      equivalence     (rpix, spix)
      character*80    oimage, nimage, errmsg
      integer         ncols, nlines, nbands, j, k, oim, nim
      integer         ier, axlen(7), naxis, pixtype, nargs, maxp
      DATA TRANS/0.,1.,0.,0.,0.,1./
      maxp=2048
      oimage="WFIcuts"
      call imopen (oimage, 1, oim, ier)   ! open the image
      if (ier .ne. 0) goto 91
      call imgsiz (oim, axlen, naxis, pixtype, ier) !! --- Determine the size and pixel type of the image being copied.
      if (ier .ne. 0) goto 91
      ncols  = axlen(1)
      nlines = axlen(2)
      nbands = axlen(3)
      write(*,*)ncols,nlines,nbands,pixtype
```

The screen output is:
```
       251           0         251           6
```
That second number should be 251 and not zero (that is the image size is 251 by 251).  The rest of the code does not work properly since the image data is not properly loaded.

Does anyone have any experience with the libimfort libraries with IRAF 2.17?

Thanks,

Jerry

**Comment by olebole** [comment, marked as answer]
What changed in the last 20 years is the port of IRAF to 64 bit. This port is limited to the **ILP64** memory model which means that integer are 64 bit by default. This requires a special setting for the Fortran compilation, since the default model is to have 32-bit integers. In IRAF itself this is done by a modified **f2c** program; however if you use an external compiler (**gfortran**), you should make sure that you choose the correct memory model.

Could this the reason for the problem?

From the [code history of `imopnx.x`](https://github.com/iraf-community/iraf/blame/history/sys/imfort/imopnx.x) I would guess that **imopen()** never supported FITS files; however I may be wrong here.

For a more modern programmatic approach to observation planning, it may be worth trying out [**astroplan**](https://astroplan.readthedocs.io/), which is an Astropy affiliated Python package specifically designed for that.

**Reply by jeromeorosz** [reply]
Hi Ole,

Thanks for the response.   That was indeed the problem.  This modified code works:

        real            rpix(2048),rbright(2048,2048),sfred
        real            trans(6),xline(2),yline(2)
        integer*2       spix(4096),ofred,sbright(2048,2048)
        equivalence     rpix, spix
        character*80    oimage, nimage, errmsg
        integer         ncols, nlines, nbands, j, k, oim, nim
        integer*8       ier, axlen(7), naxis, pixtype, nargs,maxp
        integer*8       j8,k8

[snip]

        call imgsiz (oim, axlen, naxis, pixtype, ier)
        if (ier .ne. 0) goto 91
        ncols  = int(axlen(1),4)
        nlines = int(axlen(2),4)
        nbands = int(axlen(3),4)

[snip]

        if (pixtype .eq. 3_8) then
            do k = 1, ncols
                do j = 1, nlines
                    call imgs2s (oim, ofred, j, j, k, k, ier)
                    if (ier .ne. 0) goto 91
                    sbright(j,k)=ofred
                end do
             end do
        else
            do k = 1, ncols
                k8=int(k,8)
                do j = 1, nlines
                    j8=int(j,8)
                    call imgs2r (oim, rfred, j8, j8, k8, k8, ier)
                    if (ier .ne. 0) goto 91
                    rbright(j,k)=rfred
                end do
            end do
        endif


As for the finding charts themselves, I manage to get the code working with the fitsio libraries.  However, I have other codes that use libimfort that are much more complicated, and that will require more work.

By the way, this code was running on my Linux box.  When I try the same thing on my MacBook Pro with an M1 chip, I get:

ld: warning: ignoring file libimfort.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file libsys.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file libvops.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file libos.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file libf2c.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64

gfortran was installed by Homebrew, and it looks like it targets x86_64-apple-darwin21.  I guess I need to look around for a version that is targeted to macOS-arm64. 

Thanks again for the help,

Jerry

**Reply by olebole** [reply]
You could also use the script `unix/hlib/f77.sh` from IRAF, which also uses the ILP64 memory model of IRAF and this way resolves the problem you had with **imopen()**.
